metric_metadata/cached/cached.go

```diff
@@ -111,37 +160,47 @@ func (c *CachedMetricMetadataAPI) CheckHealthy() error {
 	return c.metricMetadataAPI.CheckHealthy()
 }
 
-// getCachedTagSet is a thread-safe way to get the cached data for a metric (protected by a mutex)
-func (c *CachedMetricMetadataAPI) getCachedTagSet(metricKey api.MetricKey) CachedTagSetList {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-	return c.getAllTagsCache[metricKey]
-}
-
-// setCachedTagSet is a thread-safe way to assign cached data (protected by a mutex).
-// if the given value is less recent than the stored value, nothing happens.
-func (c *CachedMetricMetadataAPI) replaceCachedTagSet(metricKey api.MetricKey, tagsets []api.TagSet, startTime time.Time) {
-	value := CachedTagSetList{
-		TagSets: tagsets,
-		Expiry:  startTime.Add(c.timeToLive),
-	}
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-	if c.getAllTagsCache[metricKey].Expiry.Before(value.Expiry) {
-		c.getAllTagsCache[metricKey] = value
+// fetchAndUpdateCachedTagSet updates the in-memory cache (asusming the update)
+// is newer than what is in the cache. The caller MUST have the lock on the item
+// in the cache already and will release that lock before returning
+func (c *CachedMetricMetadataAPI) fetchAndUpdateCachedTagSet(metricKey api.MetricKey, context metadata.Context) ([]api.TagSet, error) {
```

Comment:
I think this should take the `item` as an argument (as a `*CachedTagSetList`) instead of looking it up in the map. In both places that it's called, the `item` pointer has already been figured out.

Note: as-is, there's a race condition in this function: it accesses `getAllTagsCache` without a lock (neither caller acquires one for it). By avoiding the map access completely (by taking `item` as a parameter), we don't have to worry about this.